File: src/ImageCompression/ProcessingModules/ModuleDCT.java
package ImageCompression.ProcessingModules;


import ImageCompression.Containers.Flag;
import ImageCompression.Containers.TripleShortMatrix;
import ImageCompression.Constants.State;
import ImageCompression.Constants.TypeQuantization;
import ImageCompression.Utils.Objects.DctConvertor;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * Created by Димка on 19.09.2016.
 */
public class ModuleDCT {
    private TripleShortMatrix tripleShortMatrix;
    private DctConvertor a,b,c;
//    private TripleShortMatrix resTripleShortMatrix;
    private Flag flag;

    public ModuleDCT(TripleShortMatrix tripleShortMatrix,Flag flag) {
        this.tripleShortMatrix = tripleShortMatrix;
//        this.resTripleShortMatrix=tripleShortMatrix;
        this.flag=flag;

        if(tripleShortMatrix.getState() == State.Yenl)//new code . Does it is needed ?
            tripleShortMatrix.setState(State.YBR);

        DctConvertor.State state=(tripleShortMatrix.getState()==State.DCT)? DctConvertor.State.DCT: DctConvertor.State.ORIGIN;

        a=new DctConvertor(tripleShortMatrix.getA(), state, TypeQuantization.luminosity, flag);
        b=new DctConvertor(tripleShortMatrix.getB(), state, TypeQuantization.Chromaticity, flag);
        c=new DctConvertor(tripleShortMatrix.getC(), state, TypeQuantization.Chromaticity, flag);

    }


    @FunctionalInterface
    private interface IThraedable{
        void doFor(DctConvertor e);
    }
    public void dataProcessing(IThraedable forEach) {
        if(tripleShortMatrix.getState() ==State.Yenl)//new code . Does it is needed ?
            tripleShortMatrix.setState(State.YBR);

        forEach.doFor(a);
        forEach.doFor(b);
        forEach.doFor(c);

        State state=a.getState()== DctConvertor.State.DCT?State.DCT:State.YBR;
        tripleShortMatrix.setState(state);
        if(tripleShortMatrix.getA().length> tripleShortMatrix.getB().length&& tripleShortMatrix.getState() ==State.YBR)
            tripleShortMatrix.setState(State.Yenl);
    }
    public void dataProcessingInThreads(IThraedable forEach) {
        if(tripleShortMatrix.getState() ==State.Yenl)//new code . Does it is needed ?
            tripleShortMatrix.setState(State.YBR);

        ExecutorService executorService = Executors.newFixedThreadPool(3);
        Future[] futures=new Future[3];

        futures[0] = executorService.submit(()->forEach.doFor(a));
        futures[1] = executorService.submit(()->forEach.doFor(b));
        futures[2] = executorService.submit(()->forEach.doFor(c));

        for (Future future : futures) {
            try {
                future.get();
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (ExecutionException e) {
                e.printStackTrace();
            }
        }

        State state=a.getState()== DctConvertor.State.DCT?State.DCT:State.YBR;
        tripleShortMatrix.setState(state);
        if(tripleShortMatrix.getA().length> tripleShortMatrix.getB().length&& tripleShortMatrix.getState() ==State.YBR)
            tripleShortMatrix.setState(State.Yenl);
    }

    public TripleShortMatrix getYCbCrMatrix(boolean isMultiThreads) {
        switch (tripleShortMatrix.getState()){
            case DCT:
                if(isMultiThreads)
                    dataProcessingInThreads(DctConvertor::getMatrixOrigin);
                else
                    dataProcessing(DctConvertor::getMatrixOrigin);
                break;
        }
        if(tripleShortMatrix.getState() ==State.YBR|| tripleShortMatrix.getState() ==State.Yenl)
            return tripleShortMatrix;
        return null;
    }
    public TripleShortMatrix getDCTMatrix(boolean isMultiThreads){
        switch (tripleShortMatrix.getState()){
            case YBR:
            case Yenl:
                if(isMultiThreads)
                    dataProcessingInThreads(DctConvertor::getMatrixDct);
                else
                    dataProcessing(DctConvertor::getMatrixDct);
                break;
        }

        if(tripleShortMatrix.getState() ==State.DCT)
            return tripleShortMatrix;
        return null;
    }

    //    public TripleShortMatrix getMatrix() {
//        return tripleShortMatrix;
//    }
}

File: ImageCompressionLib/src/ImageCompressionLib/ProcessingModules/ModuleDCT.kt
package ImageCompressionLib.ProcessingModules


import ImageCompressionLib.Containers.Type.Flag
import ImageCompressionLib.Containers.TripleShortMatrix
import ImageCompressionLib.Constants.State
import ImageCompressionLib.Constants.TypeQuantization
import ImageCompressionLib.Containers.Matrix.Matrix
import ImageCompressionLib.Containers.Parameters
import ImageCompressionLib.Utils.Functions.Dct.DctUniversalAlgorithm
import ImageCompressionLib.Utils.Objects.DctConvertor

import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import java.util.concurrent.Future
import kotlin.reflect.KFunction1

/**
 * Created by Димка on 19.09.2016.
 */
class ModuleDCT(private val tripleShortMatrixOld: TripleShortMatrix) { //    private TripleShortMatrix resTripleShortMatrix;
    private val parameter:Parameters
    private val flag:Flag
    private var a: DctConvertor
    private var b: DctConvertor
    private var c: DctConvertor
    private var state:DctConvertor.State

    init {
        parameter=tripleShortMatrixOld.parameters
        flag=parameter.flag
        val dctUtils=DctUniversalAlgorithm(parameter.unitSize)
//        if (tripleShortMatrixOld.state == State.Yenl)
//        //new code . Does it is needed ?
//            tripleShortMatrixOld.state = State.YBR

        state = if (tripleShortMatrixOld.state == State.Dct) DctConvertor.State.DCT else DctConvertor.State.ORIGIN

        a = DctConvertor(tripleShortMatrixOld.a, state, TypeQuantization.luminosity,parameter,dctUtils.copy() )
        b = DctConvertor(tripleShortMatrixOld.b, state, TypeQuantization.Chromaticity, parameter,dctUtils.copy())
        c = DctConvertor(tripleShortMatrixOld.c, state, TypeQuantization.Chromaticity, parameter,dctUtils)

    }//        this.resTripleShortMatrix=tripleShortMatrixOld;



    fun dataProcessing(forEach: KFunction1<DctConvertor, Matrix<Short>>) {
//        if (tripleShortMatrixOld.state == State.Yenl)
//        //new code . Does it is needed ?
//            tripleShortMatrixOld.state = State.YBR

        forEach.invoke(a)
        forEach.invoke(b)
        forEach.invoke(c)

        processState()
    }

    fun dataProcessingInThreads(forEach: KFunction1<DctConvertor, Matrix<Short>>) {
//        if (tripleShortMatrixOld.state == State.Yenl)
//        //new code . Does it is needed ?
//            tripleShortMatrixOld.state = State.YBR

        val executorService = Executors.newFixedThreadPool(3)
        val futures = arrayOfNulls<Future<*>>(3)

        futures[0] = executorService.submit { forEach.invoke(a) }
        futures[1] = executorService.submit { forEach.invoke(b) }
        futures[2] = executorService.submit { forEach.invoke(c) }

        for (future in futures) {
            try {
                future!!.get()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } catch (e: ExecutionException) {
                e.printStackTrace()
            }

        }

        processState()
    }
    fun getYCbCrMatrix(isMultiThreads: Boolean): TripleShortMatrix {
        if(flag.isChecked(Flag.Parameter.DCT))
            return getYCbCrMatrix1(isMultiThreads)
        else {
//            setState(DctConvertor.State.ORIGIN)
//            processState()
            return tripleShortMatrixOld
        }
    }
    fun getDCTMatrix(isMultiThreads: Boolean): TripleShortMatrix {
        if(flag.isChecked(Flag.Parameter.DCT))
            return getDCTMatrix1(isMultiThreads)
        else {
//            setState(DctConvertor.State.DCT)
//            processState()
            return tripleShortMatrixOld
        }
    }

    private fun getYCbCrMatrix1(isMultiThreads: Boolean): TripleShortMatrix {
        when (state) {
            DctConvertor.State.DCT -> if (isMultiThreads)
                dataProcessingInThreads(DctConvertor::getMatrixOrigin)
            else
                dataProcessing(DctConvertor::getMatrixOrigin)
        }
        return tripleShortMatrixOld
//        return if (tripleShortMatrixOld.state == State.YBR || tripleShortMatrixOld.state == State.Yenl) tripleShortMatrixOld else  throw Exception("State not correct")
    }
    private fun getDCTMatrix1(isMultiThreads: Boolean): TripleShortMatrix {
        when (state) {
            DctConvertor.State.ORIGIN-> if (isMultiThreads)
                dataProcessingInThreads(DctConvertor::getMatrixDct)
            else
                dataProcessing(DctConvertor::getMatrixDct)
        }

        return tripleShortMatrixOld
//        return if (tripleShortMatrixOld.state == State.DCT) tripleShortMatrixOld else throw Exception("tripleShortMatrixOld.state:${tripleShortMatrixOld.state}!=DCT")
    }
    private fun processState(){
        tripleShortMatrixOld.state = if (a.state == DctConvertor.State.DCT) State.Dct else State.Origin
    }

    //    public TripleShortMatrix getMatrix() {
    //        return tripleShortMatrixOld;
    //    }
}
